src/transformStateWithClones.js

```
@@ -7,7 +7,44 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const history = [];
+  let prevState = { ...state }; // початковий стан
+
+  for (const action of actions) {
+    let stateCopy = { ...prevState };
+
+    switch (action.type) {
+      case 'addProperties':
+        if (
+          action.extraData &&
+          typeof action.extraData === 'object' &&
+          !Array.isArray(action.extraData)
+        ) {
+          stateCopy = { ...stateCopy, ...action.extraData };
+        }
+        break;
+
+      case 'removeProperties':
+        if (Array.isArray(action.keysToRemove)) {
+          for (const key of action.keysToRemove) {
+            delete stateCopy[key];
+          }
+        }
+        break;
+
+      case 'clear':
+        stateCopy = {};
+        break;
+
+      default:
+        break;
```

This violates checklist item #3: 'switch/case should always have default case for error handling'. An empty `default` case silently ignores unknown action types, which could lead to unexpected behavior. It would be better to throw an error here to indicate that an invalid action was provided.